File: cli/azd/extensions/microsoft.azd.extensions/internal/resources/languages/python/event_manager.py
import asyncio
import logging
import threading
import time
import os
import sys
import datetime
from typing import Callable, Dict, Optional, Iterator
import queue
import grpc
from azd_client import AzdClient
import event_pb2

# Get logger - the actual configuration is done in main.py
logger = logging.getLogger(__name__)

class ProjectEventArgs:
    def __init__(self, project):
        self.project = project

class ServiceEventArgs:
    def __init__(self, project, service):
        self.project = project
        self.service = service

class ServerEventOptions:
    def __init__(self, host: Optional[str] = None, language: Optional[str] = None):
        self.host = host
        self.language = language

class EventManager:
    def __init__(self, azd_client: AzdClient):
        self._azd_client = azd_client
        self._stream = None
        self._lock = threading.RLock()
        self._messages = queue.Queue()
        self._stop_flag = threading.Event()

        self._project_handlers: Dict[str, Callable[[ProjectEventArgs], asyncio.Future]] = {}
        self._service_handlers: Dict[str, Callable[[ServiceEventArgs], asyncio.Future]] = {}

        logger.info("[EventManager] Initialized")

    async def dispose(self):
        logger.info("[EventManager] Disposing event manager...")
        self._stop_flag.set()

        # Properly close the stream
        if self._stream:
            try:
                self._stream.cancel()
                logger.info("[EventManager] Stream cancelled")
            except Exception as ex:
                logger.warning(f"[EventManager] Error while closing stream: {ex}")

        logger.info("[EventManager] Event manager disposed")

    async def ensure_initialized(self):
        """Initialize the gRPC stream if not already initialized."""
        with self._lock:
            if self._stream is None:
                logger.info("[EventManager] Initializing gRPC stream...")
                try:
                    # Create the bidirectional stream - simple approach like Go
                    self._stream = self._azd_client.events.EventStream.__call__(self._request_iterator())
                    logger.info(f"[EventManager] EventStream initialized successfully: {type(self._stream)}")
                except Exception as ex:
                    logger.exception(f"[EventManager] Failed to initialize stream: {ex}")
                    raise

    def _request_iterator(self):
        """Synchronous iterator to provide messages to gRPC."""
        logger.info("[EventManager] Starting request iterator")
        while not self._stop_flag.is_set():
            try:
                # Block until a message is available or timeout
                try:
                    message = self._messages.get(block=True, timeout=0.5)
                    # Log what we're sending
                    logger.info(f"[EventManager] Iterator yielding message: {message.WhichOneof('message_type')}")
                    yield message
                    self._messages.task_done()
                except queue.Empty:
                    # Just continue the loop if no messages
                    continue
            except Exception as ex:
                if not self._stop_flag.is_set():
                    logger.exception(f"[EventManager] Error in request iterator: {ex}")
        logger.info("[EventManager] Request iterator stopped")

    async def _send_message(self, message):
        """Queue a message to be sent to the gRPC stream."""
        # Ensure stream is initialized
        await self.ensure_initialized()

        # Log what we're sending
        message_type = message.WhichOneof('message_type')
        logger.info(f"[EventManager] Queueing message: {message_type}")

        # Add to the queue
        self._messages.put(message)

        # Wait a bit for the message to be sent
        await asyncio.sleep(0.1)

    async def receive(self):
        """Start receiving messages from the stream."""
        await self.ensure_initialized()

        # Send ready event immediately
        await self.send_ready_event()

        logger.info("[EventManager] Listening for events...")
        try:
            # Process incoming messages using a standard loop like Go
            while not self._stop_flag.is_set():
                try:
                    # This will block until a message is received
                    message = await asyncio.to_thread(next, self._stream, None)
                    if message is None:
                        # End of stream
                        break

                    logger.info(f"[EventManager] Received message: {message.WhichOneof('message_type')}")

                    # Process the message
                    if message.HasField("invoke_project_handler"):
                        await self.handle_project_event(message.invoke_project_handler)
                    elif message.HasField("invoke_service_handler"):
                        await self.handle_service_event(message.invoke_service_handler)
                    else:
                        logger.warning(f"[EventManager] Unhandled message type: {message.WhichOneof('message_type')}")
                except StopIteration:
                    # End of stream reached
                    logger.info("[EventManager] End of stream reached")
                    break
                except grpc.RpcError as rpc_error:
                    # Check for expected closure cases like in Go
                    if rpc_error.code() == grpc.StatusCode.CANCELLED:
                        logger.info("[EventManager] Stream cancelled (expected)")
                        break
                    elif rpc_error.code() == grpc.StatusCode.UNAVAILABLE:
                        logger.info("[EventManager] Stream unavailable (server closed)")
                        break
                    else:
                        logger.exception(f"[EventManager] Unexpected gRPC error: {rpc_error}")
                        break
                except Exception as ex:
                    if self._stop_flag.is_set():
                        break
                    logger.exception(f"[EventManager] Error processing message: {ex}")
                    # Brief pause to avoid tight loop on errors
                    await asyncio.sleep(0.1)
        except Exception as ex:
            logger.exception(f"[EventManager] Unexpected error in receive: {ex}")
        finally:
            logger.info("[EventManager] Receive loop exiting")

    async def add_project_event_handler(self, event_name: str, handler: Callable[[ProjectEventArgs], asyncio.Future]):
        """Add a handler for project events."""
        logger.info(f"[EventManager] Adding project event handler: {event_name}")
        self._project_handlers[event_name] = handler

        # Send subscribe message - follow Go pattern exactly
        message = event_pb2.EventMessage(
            subscribe_project_event=event_pb2.SubscribeProjectEvent(
                event_names=[event_name]
            )
        )

        # Send message directly through queue
        await self._send_message(message)
        logger.info(f"[EventManager] Project event handler registered: {event_name}")

    async def add_service_event_handler(self, event_name: str, handler: Callable[[ServiceEventArgs], asyncio.Future],
                                     options: Optional[ServerEventOptions] = None):
        """Add a handler for service events."""
        options = options or ServerEventOptions()
        logger.info(f"[EventManager] Adding service event handler: {event_name} (host={options.host}, language={options.language})")
        self._service_handlers[event_name] = handler

        # Send subscribe message - follow Go pattern exactly
        message = event_pb2.EventMessage(
            subscribe_service_event=event_pb2.SubscribeServiceEvent(
                event_names=[event_name],
                host=options.host or "",
                language=options.language or ""
            )
        )

        # Send message directly through queue
        await self._send_message(message)
        logger.info(f"[EventManager] Service event handler registered: {event_name}")

    def remove_project_event_handler(self, event_name: str):
        """Remove a project event handler."""
        logger.info(f"[EventManager] Removing project handler: {event_name}")
        self._project_handlers.pop(event_name, None)

    def remove_service_event_handler(self, event_name: str):
        """Remove a service event handler."""
        logger.info(f"[EventManager] Removing service handler: {event_name}")
        self._service_handlers.pop(event_name, None)

    async def send_ready_event(self):
        """Send the ready event to signal the extension is ready to receive events."""
        logger.info("[EventManager] Sending ExtensionReadyEvent")

        # Create ready event exactly as Go does
        message = event_pb2.EventMessage(
            extension_ready_event=event_pb2.ExtensionReadyEvent(
                status="ready"
            )
        )

        # Send message and log success
        await self._send_message(message)
        logger.info("[EventManager] ExtensionReadyEvent sent")

    async def send_project_handler_status(self, event_name: str, status: str, message: str):
        """Send status of project event handling."""
        logger.info(f"[EventManager] Sending ProjectHandlerStatus: {event_name} => {status}")

        # Create status message like Go
        status_message = event_pb2.EventMessage(
            project_handler_status=event_pb2.ProjectHandlerStatus(
                event_name=event_name,
                status=status,
                message=message
            )
        )

        # Send it to the stream
        await self._send_message(status_message)

    async def send_service_handler_status(self, event_name: str, service_name: str, status: str, message: str):
        """Send status of service event handling."""
        logger.info(f"[EventManager] Sending ServiceHandlerStatus: {event_name}/{service_name} => {status}")

        # Create status message like Go
        status_message = event_pb2.EventMessage(
            service_handler_status=event_pb2.ServiceHandlerStatus(
                event_name=event_name,
                service_name=service_name,
                status=status,
                message=message
            )
        )

        # Send it to the stream
        await self._send_message(status_message)

    async def handle_project_event(self, invoke_msg):
        """Handle a project event from the server."""
        event_name = invoke_msg.event_name
        logger.info(f"[EventManager] Handling project event: {event_name}")
        handler = self._project_handlers.get(event_name)
        status, message = "completed", ""

        if handler:
            try:
                await handler(ProjectEventArgs(invoke_msg.project))
            except Exception as ex:
                status = "failed"
                message = str(ex)
                logger.exception(f"[ProjectHandler] Error: {ex}")
            await self.send_project_handler_status(event_name, status, message)
        else:
            logger.warning(f"[EventManager] No project handler registered for event: {event_name}")

    async def handle_service_event(self, invoke_msg):
        """Handle a service event from the server."""
        event_name = invoke_msg.event_name
        logger.info(f"[EventManager] Handling service event: {event_name}")
        handler = self._service_handlers.get(event_name)
        status, message = "completed", ""

        if handler:
            try:
                await handler(ServiceEventArgs(invoke_msg.project, invoke_msg.service))
            except Exception as ex:
                status = "failed"
                message = str(ex)
                logger.exception(f"[ServiceHandler] Error: {ex}")
            await self.send_service_handler_status(event_name, invoke_msg.service.name, status, message)
        else:
            logger.warning(f"[EventManager] No service handler registered for event: {event_name}")

File: cli/azd/extensions/microsoft.azd.extensions/internal/resources/languages/javascript/eventManager.js
const grpc = require("@grpc/grpc-js");
const logger = require("./logger");
const {
  EventMessage,
  ExtensionReadyEvent,
  SubscribeProjectEvent,
  SubscribeServiceEvent,
  ProjectHandlerStatus,
  ServiceHandlerStatus,
} = require("./generated/proto/event_pb");

class EventManager {
  constructor(client) {
    this._client = client;
    this._stream = null;
    this._projectHandlers = {};
    this._serviceHandlers = {};
  }

  async close() {
    if (this._stream) {
      logger.info("Closing event stream");
      return new Promise((resolve) => {
        this._stream.end();
        resolve();
      });
    }
    return Promise.resolve();
  }

  async _ensureStream() {
    if (!this._stream) {
      logger.info("Establishing event stream");
      try {
        logger.debug("Metadata being used for stream:", {
          metadata:
            this._client && this._client._metadata
              ? this._client._metadata.getMap()
              : {},
        });

        this._stream = this._client.Events.eventStream(this._client._metadata);
        logger.info("Event stream established");
        await this._sendReadyEvent();
      } catch (err) {
        const errorMsg = `Failed to establish event stream: ${err.message}`;
        logger.error(errorMsg, { error: err });
        throw err;
      }
    }
  }

  async _sendReadyEvent() {
    const event = new EventMessage();
    const payload = new ExtensionReadyEvent();
    payload.setStatus("ready");
    payload.setMessage("JavaScript extension is initialized and ready");

    event.setExtensionReadyEvent(payload);
    this._logEvent("SEND", event.toObject());

    return new Promise((resolve, reject) => {
      this._stream.write(event, (err) => {
        if (err) {
          logger.error(`Failed to send ready event: ${err.message}`, { error: err });
          reject(err);
        } else {
          logger.info("Extension ready event sent successfully");
          resolve();
        }
      });
    });
  }

  _listen() {
    if (!this._stream) {
      logger.error("Cannot listen: stream not initialized");
      return;
    }

    this._stream.on("data", async (msg) => {
      this._logEvent("RECV", msg.toObject());

      try {
        if (msg.hasInvokeProjectHandler()) {
          await this._invokeProjectHandler(msg.getInvokeProjectHandler());
        }

        if (msg.hasInvokeServiceHandler()) {
          await this._invokeServiceHandler(msg.getInvokeServiceHandler());
        }
      } catch (err) {
        logger.error(`Error processing received message: ${err.message}`, {
          error: err,
        });
      }
    });

    this._stream.on("error", (err) => {
      if (err.code === grpc.status.UNAVAILABLE) {
        logger.warn(`Stream error (UNAVAILABLE): ${err.message}`, { error: err });
        return;
      }
      logger.error(`Stream error: ${err.message}`, { error: err });
    });

    this._stream.on("end", () => {
      logger.info("Stream ended by server (EOF)");
    });
  }

  async _invokeProjectHandler(invokeMsg) {
    const eventName = invokeMsg.getEventName();
    logger.info(`Handling project event: ${eventName}`);

    const handler = this._projectHandlers[eventName];
    if (!handler) {
      logger.warn(`No handler registered for project event: ${eventName}`);
      return;
    }

    const args = { project: invokeMsg.getProject()?.toObject() };

    let status = "completed";
    let message = "";

    try {
      await handler(args);
    } catch (err) {
      status = "failed";
      message = err.message;
      logger.error(`invokeProjectHandler error for event ${eventName}: ${err.message}`, {
        eventName,
        error: err,
      });
    }

    return this._sendProjectHandlerStatus(eventName, status, message);
  }

  async _invokeServiceHandler(invokeMsg) {
    const eventName = invokeMsg.getEventName();
    logger.info(`Handling service event: ${eventName}`);

    const handler = this._serviceHandlers[eventName];
    if (!handler) {
      logger.warn(`No handler registered for service event: ${eventName}`);
      return;
    }

    const args = {
      project: invokeMsg.getProject()?.toObject(),
      service: invokeMsg.getService()?.toObject(),
    };

    let status = "completed";
    let message = "";

    try {
      await handler(args);
    } catch (err) {
      status = "failed";
      message = err.message;
      logger.error(`invokeServiceHandler error for event ${eventName}: ${err.message}`, {
        eventName,
        serviceName: invokeMsg.getService()?.getName(),
        error: err,
      });
    }

    return this._sendServiceHandlerStatus(
      eventName,
      invokeMsg.getService()?.getName() || "",
      status,
      message
    );
  }

  _sendProjectHandlerStatus(eventName, status, message) {
    const event = new EventMessage();
    const statusMsg = new ProjectHandlerStatus();

    statusMsg.setEventName(eventName);
    statusMsg.setStatus(status);
    statusMsg.setMessage(message);

    event.setProjectHandlerStatus(statusMsg);
    this._logEvent("SEND", event.toObject());

    return new Promise((resolve, reject) => {
      this._stream.write(event, (err) => {
        if (err) {
          logger.error(`Error sending project handler status: ${err.message}`, {
            eventName,
            status,
            error: err,
          });
          reject(err);
        } else {
          resolve();
        }
      });
    });
  }

  _sendServiceHandlerStatus(eventName, serviceName, status, message) {
    const event = new EventMessage();
    const statusMsg = new ServiceHandlerStatus();

    statusMsg.setEventName(eventName);
    statusMsg.setServiceName(serviceName);
    statusMsg.setStatus(status);
    statusMsg.setMessage(message);

    event.setServiceHandlerStatus(statusMsg);
    this._logEvent("SEND", event.toObject());

    return new Promise((resolve, reject) => {
      this._stream.write(event, (err) => {
        if (err) {
          logger.error(`Error sending service handler status: ${err.message}`, {
            eventName,
            serviceName,
            status,
            error: err,
          });
          reject(err);
        } else {
          resolve();
        }
      });
    });
  }

  async receive() {
    try {
      await this._ensureStream();
      this._listen();
      logger.info("Started receiving events");
    } catch (err) {
      logger.error(`Failed to start receiving events: ${err.message}`, { error: err });
      throw err;
    }
  }

  async addProjectEventHandler(eventName, handler) {
    await this._ensureStream();
    this._projectHandlers[eventName] = handler;

    const event = new EventMessage();
    const sub = new SubscribeProjectEvent();
    sub.setEventNamesList([eventName]);
    event.setSubscribeProjectEvent(sub);

    this._logEvent("SEND", event.toObject());

    return new Promise((resolve, reject) => {
      this._stream.write(event, (err) => {
        if (err) {
          logger.error(`Error subscribing to project event ${eventName}: ${err.message}`, {
            eventName,
            error: err,
          });
          reject(err);
        } else {
          logger.info(`Subscribed to project event: ${eventName}`);
          resolve();
        }
      });
    });
  }

  async addServiceEventHandler(eventName, handler, options = {}) {
    await this._ensureStream();
    this._serviceHandlers[eventName] = handler;

    const event = new EventMessage();
    const sub = new SubscribeServiceEvent();
    sub.setEventNamesList([eventName]);
    sub.setHost(options.host || "");
    sub.setLanguage(options.language || "");
    event.setSubscribeServiceEvent(sub);

    this._logEvent("SEND", event.toObject());

    return new Promise((resolve, reject) => {
      this._stream.write(event, (err) => {
        if (err) {
          logger.error(`Error subscribing to service event ${eventName}: ${err.message}`, {
            eventName,
            options,
            error: err,
          });
          reject(err);
        } else {
          logger.info(`Subscribed to service event: ${eventName}`);
          resolve();
        }
      });
    });
  }

  removeProjectEventHandler(eventName) {
    logger.info(`Removing project event handler: ${eventName}`);
    delete this._projectHandlers[eventName];
  }

  removeServiceEventHandler(eventName) {
    logger.info(`Removing service event handler: ${eventName}`);
    delete this._serviceHandlers[eventName];
  }

  _logEvent(direction, eventObj) {
    const entries = Object.entries(eventObj);
    if (entries.length === 0) {
      logger.warn(`${direction} Empty event object`);
      return;
    }

    const [eventType, eventData] = entries[0];
    const typeName = eventType
      .split("_")
      .map((word) => word.charAt(0).toUpperCase() + word.slice(1))
      .join("");

    let details = "";
    try {
      const detailFields = [
        "event_name",
        "status",
        "event_names",
        "name",
        "service_name",
      ];
      const extracted = [];

      for (const field of detailFields) {
        if (eventData[field]) {
          if (Array.isArray(eventData[field])) {
            extracted.push(`${field}: ${eventData[field].join(", ")}`);
          } else {
            extracted.push(`${field}: ${eventData[field]}`);
          }
        }
      }

      if (eventData.service?.name) {
        extracted.push(`service: ${eventData.service.name}`);
      }

      details = extracted.join(", ");
    } catch (err) {
      details = "[Error extracting details]";
      logger.error("Error extracting event details", { error: err });
    }

    logger.info(`${direction} ${typeName} - ${details}`);
    logger.debug(`${direction} ${typeName} DETAILS`, {
      direction,
      typeName,
      eventData,
    });
  }
}

module.exports = { EventManager };
